draft-irtf-cfrg-bbs-signatures.md

```diff
@@ -403,15 +403,16 @@ Precomputations:
 
 Procedure:
 
-1. domain = calculate_domain(PK, Q_1, Q_2, (H_1, ..., H_L), header)
-2. if domain is INVALID, return INVALID
-3. e_s_for_hash = serialize((SK, domain, msg_1, ..., msg_L))
-4. if e_s_for_hash is INVALID, return INVALID
-5. (e, s) = hash_to_scalar(e_s_for_hash, 2)
-6. if e or s is INVALID, return INVALID
-7. B = P1 + Q_1 * s + Q_2 * domain + H_1 * msg_1 + ... + H_L * msg_L
-8. A = B * (1 / (SK + e))
-9. return signature_to_octets(A, e, s)
+1.  domain = calculate_domain(PK, Q_1, Q_2, (H_1, ..., H_L), header)
+2.  if domain is INVALID, return INVALID
+3.  e_s_octs = serialize((SK, domain, msg_1, ..., msg_L))
+4.  if e_s_octs is INVALID, return INVALID
+5.  e = hash_to_scalar(e_s_octs || I2OSP(0, 1))
+6.  s = hash_to_scalar(e_s_octs || I2OSP(1, 1))
+7.  if e or s is INVALID, return INVALID
```

Comment:
In this, the inputs for `e` and `s` differ in one bit only. This will highly correlate the 2 values. Maybe we can do something like,

```
5.  e = hash_to_scalar(e_s_octs || I2OSP(0, 1))
6.  if e is INVALID, return INVALID
7.  s = hash_to_scalar(I2OSP(e) || I2OSP(1, 1))
8.  if s is INVALID, return INVALID
```
This has another small issue, namely that the 2 inputs do not have the same length, which will make it hard to guarantee that the Hamiltonian distance between them is infarct large.

We could argue that since the SK is unknown, in the RO model the 2 inputs will be at least 32 random bytes apart. However ideally we would do something like:
```
5.  v = hash_to_scalar(e_s_octs || I2OSP(0, 1))
6.  if v is INVALID, return INVALID
7.  e = hash_to_scalar(I2OSP(v) || I2OSP(1, 1))
8.  if e is INVALID, return INVALID
9.  s = hash_to_scalar(I2OSP(e) || I2OSP(2, 1))
10. if s is INVALID, return INVALID
```
Here we don't require `hash_to_scalar` to be an RO, just the computational assumptions for the underlying hash will do. But at this point we rather complicate things IMO.